File: JavaScript/1682-Longest Palindromic Subsequence II.js
/**
 * @param {string} s
 * @return {number}
 */
var longestPalindromeSubseq = function(s) {
    let res = 0;
    const n = s.length;
    const memo = new Array(n).fill().map(x=> new Array(n).fill());
    
    let dp = (i, j) =>{
        if (memo[i][j] !== undefined){
            return memo[i][j];
        }
        let ans = 0;
        for (let start=i+1; start<=j-2; start++){
            for (let end=j-1; end > start; end--){
                if (s[start] === s[end] && s[start] !== s[i]){
                    ans = Math.max(ans, 2+dp(start, end));
                    break;
                }
            }
        }
        memo[i][j] = ans;
        return ans;
    }
    
    for (let i=0; i<n-1; i++){
        for (let j=n-1; j>i; j--){
            if (s[i] === s[j]){
                res = Math.max(res, 2+dp(i, j));
                break;
            }
        }
    }
    return res;
};

File: 1682-Longest Palindromic Subsequence II.py
class Solution(object):
    def longestPalindromeSubseq(self, s):
        """
        :type s: str
        :rtype: int
        """
# Time: O(n^2)
# Space: O(n^2)
        res = 0
        memo = {}
        n = len(s)
        
        def dp(i, j):
            if (i, j) in memo:
                return memo[(i,j)]
            res = 0
            for start in range(i+1, j):
                for end in range(j-1, start, -1):
                    if s[start] == s[end] and s[i] != s[start]:
                        res = max(res, 2 + dp(start, end))
                        break
            memo[(i, j)] = res
            return res
        
        for start in range(n-1):
            for end in range(n-1, start, -1):
                if s[start] == s[end]:
                    res = max(res, 2 + dp(start, end))
                    break
        return res